spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,310 @@
+
+      fileScanTasks.addAll(latestMicroBatch.tasks());
+    } while (latestMicroBatch.snapshotId() != endOffset.snapshotId());
+
+    return fileScanTasks;
+  }
+
+  private StreamingOffset getNextAvailableSnapshot(long snapshotId) {
+    Snapshot previousSnapshot = table.snapshot(snapshotId);
+    Snapshot pointer = table.currentSnapshot();
+    while (pointer != null && previousSnapshot.snapshotId() != pointer.parentId()) {
+      Preconditions.checkState(pointer.operation().equals(DataOperations.APPEND),
+          "Encountered Snapshot DataOperation other than APPEND.");
+
+      pointer = table.snapshot(pointer.parentId());
+    }
+
+    Preconditions.checkState(pointer != null,
+        "Cannot read data from snapshot which has already expired: %s", snapshotId);
+
+    return new StreamingOffset(pointer.snapshotId(), 0L, false);
+  }
+
+  interface InitialOffsetStore {
+    static InitialOffsetStore getInstance(Table table, String checkpointLocation) {
+      return new InitialOffsetStoreImpl(table, checkpointLocation);
+    }
+
+    StreamingOffset addInitialOffset();
+
+    boolean isOffsetStoreInitialized();
+
+    StreamingOffset getInitialOffset();
+  }
+
+  private static class InitialOffsetStoreImpl implements InitialOffsetStore {
+    private final Table table;
+    private final FileIO io;
+    private final String initialOffsetLocation;
+
+    InitialOffsetStoreImpl(Table table, String checkpointLocation) {
+      this.table = table;
+      this.io = table.io();
+      this.initialOffsetLocation = new Path(checkpointLocation, "offsets/0").toString();
+    }
+
+    @Override
+    public StreamingOffset addInitialOffset() {
+      table.refresh();
+      StreamingOffset offset = table.currentSnapshot() == null ?
+          StreamingOffset.START_OFFSET :
+          new StreamingOffset(SnapshotUtil.oldestSnapshot(table).snapshotId(), 0, true);
+
+      OutputFile file = io.newOutputFile(initialOffsetLocation);
+      serialize(offset, file);
+
+      return offset;
+    }
+
+    @Override
+    public boolean isOffsetStoreInitialized() {
+      InputFile file = io.newInputFile(initialOffsetLocation);
+      return file.exists();
+    }
+
+    @Override
+    public StreamingOffset getInitialOffset() {
+      InputFile file = io.newInputFile(initialOffsetLocation);
+      return deserialize(file);
+    }
+
+    private void serialize(StreamingOffset metadata, OutputFile file) {
+      try (OutputStream outputStream = file.create()) {
+        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+        writer.write(metadata.json());
+        writer.flush();
+      } catch (IOException ioException) {
+        throw new IllegalArgumentException(
+            String.format("Failed to serialize latest checkpoint to: %s, with error: %s.",
+                initialOffsetLocation, ioException));
+      }
+    }
+
+    private StreamingOffset deserialize(InputFile file) {
```

Comment:
I think that `readOffset` and `writeOffset` are more accurate names. I tend to think of `serialize` as conversion to a byte stream, not necessarily writing to durable storage.